sklearn/linear_model/tests/test_sgd.py

```diff
@@ -387,8 +387,10 @@ def test_validation_set_not_used_for_training(klass):
         cv = StratifiedShuffleSplit(test_size=validation_fraction,
                                     random_state=seed)
     else:
+        rng = np.random.RandomState(seed)
+        rng.randint(1, np.iinfo(np.int32).max)
         cv = ShuffleSplit(test_size=validation_fraction,
-                          random_state=seed)
+                          random_state=rng)
```

Comment:
Sorry I had not read the description of the PR...

I think we need to rework the test to make it less hackish.

Reply:
You can probably just move the creation of the validation split `validation_score_cb` just after the call to `check_random_state`.

It's still dependent a bit on internal details but the test would appear less convoluted.

Reply:
That's what I explain in the PR message. 


> #### Any other comments?
> 
> I had to adapt the existing unitest `test_validation_set_not_used_for_training` in file `test_sgd.py` :
> 
> With this change, the random state is "used" one time before the following line :
> 
> ```python
> validation_mask = self._make_validation_split(y)
> ```
> 
> So, I reproduce this effect in the unitest with those changes :
> 
> ```python
>         rng = np.random.RandomState(seed)
>         rng.randint(1, np.iinfo(np.int32).max)
>         cv = ShuffleSplit(test_size=validation_fraction,
>                           random_state=rng)
> ```
> 
> I am not totally satisfied, what do you think about ?
> 
> If you're OK, I add text in **what news** doc and add a non-regression test to check that the model is fully deterministic when seeding the random state like said @ogrisel
> 
> Thanks in advance for responses !



If you use the seed, the results will be different because inside `_fit_regressor` method, the random object is **run** a first time with the given `seed` before the use of `_make_validation_split`. That is why in the test, in order to reproduce that behavior, I create `rng` which is **run** a first time before the validaiton split. But I am not really convinced. 

Sorry if I am not clear !

Reply:
The **run** term is probably wrong but means in this case : 

```python
rng.randint(1, np.iinfo(np.int32).max)
```

Reply:
Alright I get it now...

Still I don't like the way this test is written. I think we should rewrite it completely. For instance by fitting a model on a random target, with and without early stopping on a validation split. The accuracy measured on `(X_train, y_train)` and the value of the `n_iter_` attribute of the model without validation set early stopping should be much higher than the model with early stopping.

Reply:
I made it such that there's no change in the model. I explained below that SGD regressors have no use of the seed attribute of the generated dataset. So I think it's better to just document that in the PR and leave this test as is. In addition the test you propose is not guaranteed, but only quite probable.